10_cdev/chardev.c

```diff
@@ -0,0 +1,366 @@
+// SPDX-License-Identifier: MIT and GPL
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/kernel.h>
+#include <linux/proc_fs.h>
+#include <linux/fs.h>
+#include <linux/err.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+
+#define  DEVICE_NAME "char_dev"
+#define  CLASS_NAME  "char_class"
+#define  LENGTH_DEF	1024
+#define	 PROC_SYS_BUF	16
+#define  PROC_DIRECTORY  "proc_chdev"
+#define  PROC_FILENAME_1  "buf_volume"
+#define  PROC_FILENAME_2  "buf_size"
+#define  SYS_FILE_NAME	  "sys_char"
+#define CLASS_ATTR(_name, _mode, _show, _store)	\
+struct class_attribute class_attr_##_name =	\
+__ATTR(_name, _mode, _show, _store)
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Anton Kotsiubailo");
+MODULE_DESCRIPTION("Task10 chardev with sysfs and procfs");
+MODULE_VERSION("0.1");
+
+static int    majorNumber;
+static char *message;
+static char *size_buffer;
+static char *volume_buffer;
+static short  size_of_message;
+static int numberOpens;
+static struct class *charClass;
+static struct device *charDevice;
+static unsigned int length = LENGTH_DEF;
+static struct proc_dir_entry *proc_dir;
+static struct proc_dir_entry *proc_file;
+static struct proc_dir_entry *proc_file2;
+module_param(length, int, 0);
+
+static int     dev_open(struct inode *, struct file *);
+static int     dev_release(struct inode *, struct file *);
+static ssize_t dev_read(struct file *, char *, size_t, loff_t *);
+static ssize_t dev_write(struct file *, const char *, size_t, loff_t *);
+static int volume_read(struct file *file_p, char __user *buffer, size_t length,
+		     loff_t *offset);
+static int size_read(struct file *file_p, char __user *buffer, size_t length,
+		     loff_t *offset);
+
+static struct file_operations volume_fops = {
+	.read = volume_read,
+};
+
+static struct file_operations size_fops = {
+	.read = size_read,
+};
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
```

Comment:
Checkpatch warning: these structs should be const!
![image](https://user-images.githubusercontent.com/79188607/112766323-ebd32e80-9019-11eb-8fdc-b76870613c00.png)